dev, v3.0: add the split region doc (#1332)
pingcap/docs: dev/reference/sql/statements/split-region.md

```
region1  [ minIndexValue               , ("a", "2000-01-01 00:00:01"))
region2  [("a", "2000-01-01 00:00:01") , ("b", "2019-04-17 14:26:19"))
region3  [("b", "2019-04-17 14:26:19") , ("c", "")                   )
region4  [("c", "")                    , maxIndexValue               )
```

## pre_split_regions

When using table with `shard_row_id_bits`, it is suggested to use with `pre_split_regions` if you want to spilt Region evenly at the start of building table. `pre_split_regions` spilt `2^(pre_split_regions-1)` Region evenly after the table is built.

> **Note:**
>
> `pre_split_regions` must amount to less than or equal to `shard_row_id_bits`.

### Example

{{< copyable "sql" >}}

```sql
create table t (a int, b int,index idx1(a)) shard_row_id_bits = 4 pre_split_regions=3;
```

After building that table, this statement will cut out 4 + 1  Regions form table t. 4 (2^(3-1)) Regions for saving table row data,1 Region for saving idx1 index data.

The range of 4 table Region are shown as follows:

```
region1:   [ -inf      ,  1<<61 )  
region2:   [ 1<<61     ,  2<<61 )
region3:   [ 2<<61     ,  3<<61 )
region4:   [ 3<<61     ,  +inf  )
```

> Why the amount of Region is 2^(pre_split_regions-1)? Because when using shard_row_id_bits, only positive numbers will be assigned to `_tidb_rowid`, so there is no need to do Spilt Region for the negative interval.

Reviewer: ```suggestion
> **Note:**
>
> The amount of split Region is 2^(pre_split_regions-1) because when using shard_row_id_bits, only positive numbers will be assigned to `_tidb_rowid`, so there is no need to do Spilt Region for the negative range.
```

Reviewer: I think we better make this a Note.

## Related session variable

> There are two `SPLIT REGION` related session variables: `tidb_wait_split_region_finish` and `tidb_wait_split_region_timeout`. See[TiDB specific system variables and syntax](/dev/reference/configuration/tidb-server/tidb-specific-variables.md).

Reviewer: ```suggestion
There are two `SPLIT REGION` related session variables: `tidb_wait_split_region_finish` and `tidb_wait_split_region_timeout`. For details, see [TiDB specific system variables and syntax](/reference/configuration/tidb-server/tidb-specific-variables.md).
```

---
title: Split Region
summary: An overview of the usage of Split Region for the TiDB database.
category: reference
---

# Split Region

For each new table created in TiDB, one Region is segmented by default to store the data of this table. This default behavior is controlled by `split-table` in the configuration file. When the data in this Region exceeds the default Region size limit, the Region starts to split into two.

> In the above case, because there is only one Region at the beginning,  all write requests occur on the TiKV where the Region is located. If there are a large number of writes for the newly created table, hotspots are caused. 

Reviewer: ```suggestion
In the above case, because there is only one Region at the beginning, all write requests occur on the TiKV where the Region is located. If there are a large number of writes for the newly created table, hotspots are caused. 
```

To solve the hotspot problem in the above scenario, TiDB introduces the pre-split function, which can pre-split multiple Regions for a certain table according to the specified parameters and scatter them to each TiKV node.

## Usage of Split Region 

There are two types of Split Region syntax:

{{< copyable "sql" >}}

```sql
SPLIT TABLE table_name [INDEX index_name] BETWEEN (lower_value) AND (upper_value) REGIONS region_num
```

`BETWEEN lower_value AND upper_value REGIONS region_num` defines the upper boundary, the lower boundary, and the Region amount. Then the current region will be evenly spilt into the number of regions (as specified in `region_num`) between the upper and lower boundaries.

{{< copyable "sql" >}}

```sql
SPLIT TABLE table_name [INDEX index_name] BY (value_list) [, (value_list)] ...
```

> `BY value_list…` is used to specify a series of points manually, then the current Region will be spilt according to them. It is suitable for scenarios with unevenly distributed data.

Reviewer: ```suggestion
`BY value_list…` is used to specify a series of points manually, then the current Region will be spilt according to them. It is suitable for scenarios with unevenly distributed data.
```
```suggestion
`BY value_list…` specifies a series of points manually, based on which the current Region is spilt. It is suitable for scenarios with unevenly distributed data.
```

### Split Table Region

The key of row data in each table is encoded by `table_id` and `row_id`. The format is as follows:

```go
t[table_id]_r[row_id]
```

For example, when `table_id` is 22 and `row_id` is 11:

```go
t22_r11
```

> Row data in the same table have the same `table_id`, but each has its unique `row_id`. This could be used for Region split.

Reviewer: ```suggestion
Row data in the same table have the same `table_id`, but each row has its unique `row_id` that can be used for Region split.
```

#### Even Split

Because `row_id` is an integer, the value of the key to be split can be calculated easily according to the `lower_value`, `upper_value`, and `region_num` defined by the user. The calculation starts with the step value (`step = (upper_value - lower_value)/num`). Then split will be done evenly per each "step" between `lower_value` and `upper_value` to generate the number of Regions as specified by `num`.

For example, if you want 16 evenly split Regions split from key range`minInt64`~`maxInt64` for table t, you can use this statement:

{{< copyable "sql" >}}

```sql
SPLIT TABLE t BETWEEN (-9223372036854775808) AND (9223372036854775807) REGIONS 16;
```

> This statement splits table t into 16 Regions between minInt64 and maxInt64. If the given primary key range is smaller than the specified one, for example, 0~1000000000,  You can use 0 and 1000000000 take place of minInt64 and maxInt64 respectively to split Regions.

Reviewer 2: ```suggestion
This statement splits table t into 16 Regions between minInt64 and maxInt64. If the given primary key range is smaller than the specified one, for example, 0~1000000000, you can use 0 and 1000000000 take place of minInt64 and maxInt64 respectively to split Regions.
```

{{< copyable "sql" >}}

```sql
SPLIT TABLE t BETWEEN (0) AND (1000000000) REGIONS 16;
```

#### Uneven split

If the known data is unevenly distributed, and you want a Region to be split respectively in key ranges -inf ~ 10000, 10000 ~ 90000, and 90000 ~ +inf, you can achieve this by setting fixed points, as shown below:

{{< copyable "sql" >}}

```sql
SPLIT TABLE t BY (10000), (90000);
```

### Split index Region

The key of the index data in the table is encoded by `table_id`, `index_id`, and the value of the index column. The format is as follows:

```go
t[table_id]_i[index_id][index_value]
```

For example, when `table_id` is 22, `index_id` is 5, and `index_value` is abc:

```go
t22_i5abc
```

The `table_id` and `index_id` of the same index data in one table is the same. To split index Regions, you need to split Regions based on `index_value`.

#### Even Spilt

The way to split index evenly works the same as splitting data evenly. However, calculating the value of step is more complicated, because `index_value` might not be an integer.

> The value of `upper` and `lower` will be encoded into a byte array firstly. After removing the longest common prefix of `lower` and `upper` byte array, the first 8 bytes of `lower` and `upper` will be converted into the uint64 format. Then `step = (upper - lower)/num` will be calculated. After that, the calculated step will be encoded into a byte array, which will be appended to the longest common prefix of the `lower` and `upper` byte array for index split. Here is an example:

Reviewer 2: ```suggestion
The values of `upper` and `lower` are encoded into a byte array firstly. After removing the longest common prefix of `lower` and `upper` byte array, the first 8 bytes of `lower` and `upper` are converted into the uint64 format. Then `step = (upper - lower)/num` is calculated. After that, the calculated step is encoded into a byte array, which is appended to the longest common prefix of the `lower` and `upper` byte array for index split. Here is an example:
```

> If the row of index idx is an integer, you can use the following SQL statement to split index data:

Reviewer 2: ```suggestion
If the column of the `idx` index is of the integer type, you can use the following SQL statement to split index data:
```

{{< copyable "sql" >}}

```sql
SPLIT TABLE t INDEX idx BETWEEN (-9223372036854775808) AND (9223372036854775807) REGIONS 16;
```

This statement splits the Region of index idx in table t into 16 Regions from `minInt64` to `maxInt64`.

If the column of index idx1 is of varchar type, and you want to split index data by prefix letters.

{{< copyable "sql" >}}

```sql
SPLIT TABLE t INDEX idx1 BETWEEN ("a") AND ("z") REGIONS 26;
```

> This statement splits index idx1 into 26 Regions from a~z. The range of region 1 is [minIndexValue, b); the range of region 2 is [b, c); … the range of region26 is [z, minIndexValue]. For index idx, data with prefix a is written into region 1; data with prefix b will be written into region 2, and so on.

Reviewer 2: ```suggestion
This statement splits index idx1 into 26 Regions from a~z. The range of Region 1 is `[minIndexValue, b)`; the range of Region 2 is `[b, c)`; … the range of Region 26 is `[z, minIndexValue]`. For the `idx` index, data with the `a` prefix is written into Region 1, while data with the `b` prefix is written into Region 2, and so on.
```

For data Region split of joint indexes, the only difference is that you can specify multiple columns values.

For example, index `idx3 (a, b)` contains 2 columns, with column `a` of timestamp type and column `b` int. If you just want to do a time range split according to column `a`, you can use the SQL statement for splitting time index of a single column. In this case, do not specify the value of column `b` in `lower_value` and `upper_velue`.

{{< copyable "sql" >}}

```sql
SPLIT TABLE t INDEX idx3 BETWEEN ("2010-01-01 00:00:00") AND ("2020-01-01 00:00:00") REGIONS 10;
```

Within the same range of time, if you want to do one more split according to column b column. Just specify the value for column b when splitting.

{{< copyable "sql" >}}

```sql
SPLIT TABLE t INDEX idx3 BETWEEN ("2010-01-01 00:00:00", "a") AND ("2010-01-01 00:00:00", "z") REGIONS 10;
```

This statement splits 10 Regions in the range of a~z according to the value of column b, with the same time prefix as column a. If the value specified for column a is different, the value of column b might not be used in this case.

#### Uneven Split

Index data can also be split by specified index values.

> For example, there is idx4 (a,b), with column a of varchar type and column b timestamp.

Reviewer 2: ```suggestion
For example, there is `idx4 (a,b)`, with column `a` of the varchar type and column `b` of the timestamp type.
```

{{< copyable "sql" >}}

```sql
SPLIT TABLE t1 INDEX idx4 ("a", "2000-01-01 00:00:01"), ("b", "2019-04-17 14:26:19"), ("c", "");  
```

> This statement specifies 3 values to split 4 Regions. The range of each Region are shown as follows:

Reviewer 2: ```suggestion
This statement specifies 3 values to split 4 Regions. The range of each Region is as follows:
```

```
region1  [ minIndexValue               , ("a", "2000-01-01 00:00:01"))
region2  [("a", "2000-01-01 00:00:01") , ("b", "2019-04-17 14:26:19"))
region3  [("b", "2019-04-17 14:26:19") , ("c", "")                   )
region4  [("c", "")                    , maxIndexValue               )
```

## pre_split_regions

> To have evenly split Regions when a table is created, it is recommended you use `shard_row_id_bits` together with `pre_split_regions`.  When a table is created successfully, `pre_split_regions` pre-spilts tables into the number of Regions as specified by `2^(pre_split_regions-1)`.

Reviewer 2: ```suggestion
To have evenly split Regions when a table is created, it is recommended you use `shard_row_id_bits` together with `pre_split_regions`. When a table is created successfully, `pre_split_regions` pre-spilts tables into the number of Regions as specified by `2^(pre_split_regions-1)`.
```

> **Note:**
>
> The value of `pre_split_regions` must be less than or equal to that of `shard_row_id_bits`.

### Example

{{< copyable "sql" >}}

```sql
create table t (a int, b int,index idx1(a)) shard_row_id_bits = 4 pre_split_regions=3;
```

> After building the table, this statement splits 4 + 1  Regions for table t. 4 (2^(3-1)) Regions are used to save table row data, and 1 Region is for saving the index data of idx1.

Reviewer 2: ```suggestion
After building the table, this statement splits 4 + 1 Regions for table t. `4 (2^(3-1))` Regions are used to save table row data, and 1 Region is for saving the index data of `idx1`.
```

> The ranges of the 4 table Regions are shown as follows:

Reviewer 2: ```suggestion
The ranges of the 4 table Regions are as follows:
```
